src/protocol/sort/reshare.rs

```diff
@@ -0,0 +1,138 @@
+        let (r0, r1) = prss.generate_fields(record_id.into());
+
+        let inputs = self.input.as_tuple();
+        // `to_helper.left` calculates part1 = (input.0 + input.1) - r1 and sends part1 to `to_helper.right`
+        if channel.identity() == to_helper.peer(Direction::Left) {
+            let part1 = inputs.0 + inputs.1 - r1;
+            channel
+                .send(
+                    to_helper.peer(Direction::Right),
+                    record_id,
+                    PartValue { part: part1 },
+                )
+                .await?;
+
+            // Sleep until `to_helper.left` sends us their part2 value
+            let PartValue { part: part2 } = channel
+                .receive(channel.identity().peer(Direction::Left), record_id)
+                .await?;
+
+            Ok(Replicated::new(part1 + part2, r1))
+        } else if channel.identity() == to_helper.peer(Direction::Right) {
+            //  `to_helper.right` calculates part2 = (input.0 - r0) and sends it to `to_helper.left`
+            let part2 = inputs.0 - r0;
+            channel
+                .send(
+                    to_helper.peer(Direction::Left),
+                    record_id,
+                    PartValue { part: part2 },
+                )
+                .await?;
+
+            // Sleep until `to_helper.right` sends us their part1 value
+            let PartValue::<F> { part: part1 } = channel
+                .receive(channel.identity().peer(Direction::Right), record_id)
+                .await?;
+
+            Ok(Replicated::new(r0, part1 + part2))
+        } else {
+            Ok(Replicated::new(r0, r1))
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::rngs::mock::StepRng;
+    use tokio::try_join;
+
+    use crate::{
+        field::Fp31,
+        helpers::Identity,
+        protocol::{sort::reshare::Reshare, QueryId, RecordId},
+        test_fixture::{
+            make_contexts, make_world, share, validate_and_reconstruct, TestStep, TestWorld,
+        },
+    };
+
+    #[tokio::test]
+    pub async fn reshare() {
+        let mut rand = StepRng::new(100, 1);
+
+        let input = Fp31::from(23_u128);
+        let share = share(input, &mut rand);
+        let record_id = RecordId::from(1);
+
+        let world: TestWorld<TestStep> = make_world(QueryId);
+        let context = make_contexts(&world);
+
+        let step = TestStep::Reshare(1);
+
+        let reshare0 = Reshare::new(share[0]);
+        let reshare1 = Reshare::new(share[1]);
+        let reshare2 = Reshare::new(share[2]);
+
+        let h0_future = reshare0.execute(&context[0], record_id, step, Identity::H2);
+        let h1_future = reshare1.execute(&context[1], record_id, step, Identity::H2);
+        let h2_future = reshare2.execute(&context[2], record_id, step, Identity::H2);
+
+        let f = try_join!(h0_future, h1_future, h2_future).unwrap();
+        let output_share = validate_and_reconstruct(f);
+        assert_eq!(output_share, input);
```

Comment:
Where do we check that this is a different sharing than the original sharing?

Reply:
good point! let me add that check